### PR TITLE
Bookmark in Mentorlist and User Profile

### PR DESCRIPTION
**Pull request by SarahhhC**
* Bookmark on/off in mentorlist and user profile.
* Add bookmark list in MyPage/Bookmark.

### CODE REVIEW

File: src/components/Bookmark/Bookmark.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    if (error) {
+      alert(error);
+    } else if (result) {
+      this.onRequestSuccess(result);
+    }
+  }
+  
+  onRequestSuccess(result) {
+    let bookmarked = result;
+    bookmarked = bookmarked.map((value) => value);
```
Comment (ifndefdeadmau5) [inline src/components/Bookmark/Bookmark.js]: Isn't it assigning same thing?

File: src/components/Bookmark/Bookmark.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  componentDidMount() {
+    UserUtil.getBookmarkList(this.onRequestCallback.bind(this));
+  }
+  
+  // Receive props befofe completely changed
+  componentWillReceiveProps(props) {
+    UserUtil.getBookmarkList(this.onRequestCallback.bind(this));
+  }
+  
+  // Render loading page while fetching user profiles
```
Comment (ifndefdeadmau5) [inline src/components/Bookmark/Bookmark.js]: user profiles?

File: src/components/Bookmark/Bookmark.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        size='small'
+      />
+    );
+  }
+  
+  renderRow(rowData) {
+    return <Row dataSource={rowData} me={this.props.me}/>;
+  }
+  
+  renderConnected() {
```
Comment (ifndefdeadmau5) [inline src/components/Bookmark/Bookmark.js]: Why this function name has 'Connected'?

File: src/components/Bookmark/Row.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+      currentStatus: this.getCurrentStatus(),
+      loaded: true,
+    });
+  }
+  
+  getProfileImage() {
+    let image;
+    if (this.props.dataSource.profile_picture) {
+      image = { uri: this.props.dataSource.profile_picture_small ?
+        this.props.dataSource.profile_picture_small : this.props.dataSource.profile_picture };
```
Comment (ifndefdeadmau5) [inline src/components/Bookmark/Row.js:44]: camelCase?

File: src/components/Bookmark/Row.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    let company = this.state.company;
+    
+    const experience = this.props.dataSource.experience;
+    if (experience && experience.length > 0) {
+      company = experience[0].employer.name;
+      if (experience[0].position) {
+        position = experience[0].position.name;
+      }
+      
+      return position + ' at ' + company;
```
Comment (ifndefdeadmau5) [inline src/components/Bookmark/Row.js:71]: This function only return value in this if statement. what if condition is not true?

File: src/components/UserList/Row.js [hunk 6]
```diff
@@ -147,10 +149,28 @@ class Row extends Component {
...
+  }
 
   sendRequest() {
     Actions.requestPage({ id: this.props.dataSource._id, me: this.props.dataSource.me });
   }
+  
+  setBookmark() {
+    if (this.state.bookmarked) {
+      UserUtil.bookmarkOff(this.onRequestCallback.bind(this), this.props.dataSource._id);
+      this.state.bookmarked = false;
```
Comment (ifndefdeadmau5) [inline src/components/UserList/Row.js]: What if 'bookmarkoff' request has failed? Does bookmarked will change it's state to false?

File: src/components/UserList/Row.js [hunk 7]
```diff
@@ -147,10 +149,28 @@ class Row extends Component {
...
+    } else {
+      UserUtil.bookmarkOn(this.onRequestCallback.bind(this), this.props.dataSource._id);
+      this.state.bookmarked = true;
```
Comment (ifndefdeadmau5) [inline src/components/UserList/Row.js]: Same above

File: src/components/UserList/Row.js [hunk 8]
```diff
@@ -194,6 +214,18 @@ class Row extends Component {
         </TouchableOpacity>
       );
     }
+    
+    let bookmarkSource = null;
+    
+    if (this.state.bookmarked) {
+      bookmarkSource = (
```
Comment (ifndefdeadmau5) [inline src/components/UserList/Row.js]: How about 'bookmarkImage'?

File: src/components/userProfile/UserProfile.js [hunk 9]
```diff
@@ -24,63 +24,94 @@ import UserOverview from './UserOverview';
 import UserUtil from '../../utils/UserUtil';
 import Icon from 'react-native-vector-icons/FontAwesome';
 
+const backBtnImg = require('../../resources/icon-arrow-left-white.png')
+const backBtnScrolledImg = require('../../resources/icon-arrow-left-grey.png');
+const bookmarkImg = require('../../resources/icon-bookmark.png');
```
Comment (ifndefdeadmau5) [inline src/components/userProfile/UserProfile.js]: How about 'bookmarkEmptyImage'?

**Review by ifndefdeadmau5 — CHANGES_REQUESTED** [review]
Thank you for your hard work! Please check some comment and fix JSCS errors.

File: src/components/Bookmark/Bookmark.js (hunk 1 above)
Comment (SarahhhC) [reply src/components/Bookmark/Bookmark.js]: Result itself contains status code too, so this is for getting arrays in result.

File: src/components/Bookmark/Bookmark.js (hunk 2 above)
Comment (SarahhhC) [reply src/components/Bookmark/Bookmark.js]: Oh should be `user profiles` => `bookmark lists`

File: src/components/Bookmark/Bookmark.js (hunk 3 above)
Comment (SarahhhC) [reply src/components/Bookmark/Bookmark.js]: `renderConnected` => `renderBookmarked`

File: src/components/Bookmark/Row.js (hunk 4 above)
Comment (SarahhhC) [reply src/components/Bookmark/Row.js:44]: Already server side send variables as that name...Unfortunately...

File: src/components/Bookmark/Row.js (hunk 5 above)
Comment (SarahhhC) [reply src/components/Bookmark/Row.js:71]: Nothing shows for description.

File: src/components/UserList/Row.js (hunk 6 above)
Comment (SarahhhC) [reply src/components/UserList/Row.js]: Yes, just for image temporarily but after refresh the state would be same as before.

File: src/components/UserList/Row.js (hunk 7 above)
Comment (SarahhhC) [reply src/components/UserList/Row.js]: So should I **do not change** the image if request has failed?

File: src/components/UserList/Row.js (hunk 8 above)
Comment (SarahhhC) [reply src/components/UserList/Row.js]: Good 👍

File: src/components/userProfile/UserProfile.js (hunk 9 above)
Comment (SarahhhC) [reply src/components/userProfile/UserProfile.js]: Nice 👍

File: src/components/Bookmark/Bookmark.js (hunk 1 above)
Comment (ifndefdeadmau5) [reply src/components/Bookmark/Bookmark.js]: Oh, then...why don't you use
`
let bookmarked = result.slice()
`
?

File: src/components/Bookmark/Row.js (hunk 5 above)
Comment (ifndefdeadmau5) [reply src/components/Bookmark/Row.js:71]: How about return null for other case to reveal hidden intent?

File: src/components/Bookmark/Bookmark.js (hunk 1 above)
Comment (SarahhhC) [reply src/components/Bookmark/Bookmark.js]: After adjustment, It worked!! Thanks!!

File: src/components/UserList/Row.js (hunk 7 above)
Comment (ifndefdeadmau5) [reply src/components/UserList/Row.js]: I guess yes~?
I mean thinking of scenario,

bookmarkOn 
->this.state.bookmarked = true 
->but bookmarkOn request fail

client bookmarked state === true but, server bookmarked state === still false, right?
Shoudn't we synchronize bookmarked state? or no?

File: src/components/Bookmark/Row.js (hunk 5 above)
Comment (SarahhhC) [reply src/components/Bookmark/Row.js:71]: Good, add else statement!

File: src/components/UserList/Row.js (hunk 7 above)
Comment (SarahhhC) [reply src/components/UserList/Row.js]: Maybe synchronizing status would be better. 😄  Thank you!

**Comment by ifndefdeadmau5** [comment]
LGTM

[Approved with PullApprove](https://pullapprove.com/sv-bootcamp/bridgeme-client/pull-request/167/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=sv-bootcamp/bridgeme-client)

**Review by ifndefdeadmau5 — APPROVED** [review]

**Pull request merged**